src/evaluators/BinaryExpression.js

```diff
@@ -61,6 +61,13 @@ export function getPureBinaryOperationResultType(
     let leftPure = purityTest(realm, lval);
     let rightPure = purityTest(realm, rval);
     if (leftPure && rightPure) return typeIfPure;
+
+    if (realm.isInPureScope()) {
+      if (!leftPure) Leak.leakValue(realm, lval);
```

Comment:
Consider renaming the enclosing function to `leakOrReportErrorIfNotPure

Reply:
Done, are you happy with this PR apart from that?